File: 0118_pascals_triangle.swift
// https://leetcode.com/problems/pascals-triangle/

class Solution {
	func generate(_ numRows: Int) -> [[Int]] {
		var result = [[1]]
		var currentRow = [1]
		if numRows == 1 { return result }
		currentRow = [1, 1]
		result.append(currentRow)
		if numRows == 2 { return result }
		for _ in 3...numRows {
			var newRow = [1]
			for j in 0..<(currentRow.count - 1) {
				newRow.append(currentRow[j] + currentRow[j + 1])
			}
			newRow.append(1)
			result.append(newRow)
			currentRow = newRow
		}
		return result
	}
}

File: 0118_pascals_triangle.kt
// https://leetcode.com/problems/pascals-triangle/description/

class Solution {
    fun generate(numRows: Int): List<List<Int>> {
        if (numRows == 1) {
            return listOf(listOf(1))
        } else if (numRows == 2) {
            return listOf(listOf(1), listOf(1, 1))
        }
        val result = ArrayDeque<List<Int>>()
        result.add(listOf(1))
        result.add(listOf(1, 1))
        for (i in 3 .. numRows) {
            val lastList = result.last()
            val newList = mutableListOf(1)
            for (j in 1 until lastList.size) {
                newList.add(lastList[j - 1] + lastList[j])
            }
            newList.add(1)
            result.add(newList.toList())
        }
        return result.toList()
    }
}
